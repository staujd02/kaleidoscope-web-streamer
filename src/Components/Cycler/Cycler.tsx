import React from 'react';
import Stream from '../Stream/Stream';
import TimeDisplay from '../TimeDisplay/TimeDisplay';

class Cycler extends React.Component<CyclerProps, CyclerState> {
    
    private readonly increment = 50;

    constructor(props: any){
        super(props);
        this.getSource = this.getSource.bind(this);
        this.state = {
            activeSource: 0,
            millisecondsRemaining: this.props.cycleTime,
            forgoRender: false
        }
    }

    shouldComponentUpdate(){
        const { forgoRender } = this.state;
        return !forgoRender ;
    }
    
    cycle = (sourceList: Array<Source>, activeSource: number) => {
        this.setState({
           activeSource: this.cycleIndex(activeSource, sourceList.length),
           millisecondsRemaining: this.props.cycleTime,
           forgoRender: false
        });
    }

    tick(sourceList: Array<Source>, activeSource: number) {
        const { millisecondsRemaining } = this.state;
        let nextRemainingCycle = millisecondsRemaining - this.increment;
        if (nextRemainingCycle > this.increment)
            this.setState({
                millisecondsRemaining: nextRemainingCycle
            });
        else
            this.cycle(sourceList, activeSource);
    }

    private cycleIndex(activeSource: number, length: number) {
        activeSource++;
        if (activeSource === length) {
            activeSource = 0;
        }
        return activeSource;
    }

    timeClicked = (event: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
        this.setState({
            millisecondsRemaining: 0,
            forgoRender: true
        });
    }

    render() {
        const { activeSource, millisecondsRemaining } = this.state;
        const { sourceList } = this.props;
        setTimeout(() => this.tick(sourceList, activeSource), this.increment);
        return (
            <div>
                <Stream source={this.getSource(activeSource)} />
                <TimeDisplay onClick={this.timeClicked} time={millisecondsRemaining} />
            </div>
        );
    }

    private getSource(activeSource: number) {
        return this.props.sourceList[activeSource];
    }
}

export default Cycler;