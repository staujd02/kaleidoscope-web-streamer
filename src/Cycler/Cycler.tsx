import React from 'react';
import Stream from '../Stream/Stream';
import TimeDisplay from '../TimeDisplay/TimeDisplay';

class Cycler extends React.Component<CyclerProps, CyclerState> {

    constructor(props: any){
        super(props);
        this.getSource = this.getSource.bind(this);
        this.state = {
            activeSource: 0,
            millisecondsRemaining: this.props.cycleTime
        }
    }
    
    cycle(sourceList: Array<Source>, activeSource: number) {
        this.setState({
           activeSource: this.cycleIndex(activeSource, sourceList.length),
           millisecondsRemaining: this.props.cycleTime
        });
    }

    tick(sourceList: Array<Source>, activeSource: number) {
        const { millisecondsRemaining } = this.state;
        let nextRemainingCycle = millisecondsRemaining - 1000;
        if (nextRemainingCycle > 1000)
            this.setState({
                millisecondsRemaining: nextRemainingCycle
            });
        else
            this.cycle.bind(this)(sourceList, activeSource);
    }

    private cycleIndex(activeSource: number, length: number) {
        activeSource++;
        if (activeSource === length) {
            activeSource = 0;
        }
        return activeSource;
    } 

    render() {
        const { activeSource, millisecondsRemaining } = this.state;
        const { sourceList } = this.props;
        setTimeout(() => this.tick(sourceList, activeSource), 1000);
        return (
            <div>
                <Stream source={this.getSource(activeSource)} />
                <TimeDisplay time={millisecondsRemaining} />
            </div>
        );
    }

    private getSource(activeSource: number) {
        return this.props.sourceList[activeSource];
    }
}

export default Cycler;