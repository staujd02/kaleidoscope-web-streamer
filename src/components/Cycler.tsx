import React from 'react';
import Stream from './Stream/Stream';

class Cycler extends React.Component<CyclerProps, CyclerState> {

    constructor(props: any){
        super(props);
        this.getSource = this.getSource.bind(this);
        this.state = {
            activeSource: 0
        }
    }

    render() {
        const { activeSource } = this.state;
        return (
            <div>
                <Stream source={this.getSource(activeSource)} />
            </div>
        );
    }

    private getSource(activeSource: number) {
        return this.props.sourceList[activeSource];
    }
}

export default Cycler;