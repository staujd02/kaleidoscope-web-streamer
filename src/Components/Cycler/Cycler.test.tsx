import React from 'react';
import Cycler from './Cycler';
import { ShallowWrapper, shallow } from 'enzyme';
import { formatHTML } from '../../TestUtilities/htmlFormatter';

describe('The Cycler', () => {

    const SourceList = getSourceList() as Array<Source>;
    const CycleTime = 64000;

    let wrapper: ShallowWrapper<CyclerProps, CyclerState>;

    beforeEach(() => {
        wrapper = shallow(<Cycler sourceList={SourceList} cycleTime={CycleTime} />)
    });

    it('renders correctly', () => expect(formatHTML(wrapper.html())).toMatchSnapshot());



    function getSourceList(): Array<Source> {
        return [
            {
                source: "https://www.google.com",
                title: "Google"
            } as Source,
            {
                source: "https://www.facebook.com",
                title: "Facebook"
            } as Source
        ];
    }

});