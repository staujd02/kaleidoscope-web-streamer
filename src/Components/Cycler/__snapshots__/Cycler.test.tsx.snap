// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Cycler renders correctly 1`] = `
"
<div>
  <iframe title=\\"Google\\" allow=\\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\\" src=\\"https://www.google.com\\">Your browser doesn't support iFrames.</iframe>
  <div class=\\"time-display\\">
    <div class=\\"time-box\\">00:01</div>
  </div>
  <div id=\\"cycler-controls\\" style=\\"visibility: hidden; webkit-transform: translateY(-0px); transform: translateY(-0px);\\">
    <div class=\\"cycler-controls\\">
      <button class=\\"MuiButtonBase-root MuiIconButton-root\\" tabindex=\\"0\\" type=\\"button\\" id=\\"skip-control\\" aria-label=\\"skip\\">
        <span class=\\"MuiIconButton-label\\">
          <svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\" role=\\"presentation\\">
            <path d=\\"M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z\\"></path>
          </svg>
        </span>
        <span class=\\"MuiTouchRipple-root\\"></span>
      </button>
      <button id=\\"playback-control\\"> Pause </button>
    </div>
  </div>
</div>
"
`;
